add entities, add entity configuration, add maping file, create a rep…

CrazyTapok opened the pull request:
1. Screenshot: 
![image](https://github.com/user-attachments/assets/104e56ec-b790-485c-9d5b-5c7f2149c205)

2. What has been implemented:
  - [x] Create Entities
  - [x] Create entity configurations
  - [x] Create a repository interface
  - [x] Create generic repository
  - [x] Create business models 
  - [x] Creat AutoMapper file to map entities and business models

CrazyTapok committed:
add entities, add entity configuration, add maping file, create a repository interface, create generic repository, add business models

controlofthesea commented on src/TaskManager.Core/Enums/ERole.cs:
```diff
@@ -0,0 +1,15 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TaskManager.Core.Enums
+{
+    public enum ERole
+    {
+        Developer,
+        ProjectManeger,
+        Admin
+    }
+}
```
Вижу, что пользуешься dotnet 8, тогда можно пользоваться и его фичами :)
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/file-scoped-namespaces
Меньше скобок и отступов в файлах будет)

controlofthesea commented on src/TaskManager.Core/Interfaces/IRepository.cs:
```diff
@@ -0,0 +1,12 @@
+﻿
+namespace TaskManager.Core.Interfaces
+{
+    public interface IRepository<TEntity, TModels> where TEntity : class where TModels : class
```
`TModels -> TModel` тут же один экземпляр подразумевается

controlofthesea commented on src/TaskManager.Core/Enums/ERole.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace TaskManager.Core.Enums
+{
+    public enum ERole
```
а почему ты называешь энамы с Е?

CrazyTapok replied:
Тип как интерфейс начинается с I, я енам называю с E.
Так не делать?

controlofthesea replied:
я не видел такого никогда)

controlofthesea replied:
думаю интерфейс с I потому что всешгда есть реализация с таким же названием и это I нужна, чтобы не путать. А с энамом не знаю причин так делать)

controlofthesea commented on src/TaskManager.Infrastructure/Mappings/MappingProfile.cs:
```diff
@@ -0,0 +1,15 @@
+﻿using AutoMapper;
+
+namespace TaskManager.Infrastructure.Mappings
+{
+    public class MappingProfile : Profile
+    {
+        public MappingProfile()
+        {
+            CreateMap<Entities.Company, Core.Models.Company>().ReverseMap();
+            CreateMap<Entities.Employee, Core.Models.Employee>().ReverseMap();
+            CreateMap<Entities.Project, Core.Models.Project>().ReverseMap();
+            CreateMap<Entities.Task, Core.Models.Task>().ReverseMap();
+        }
+    }
+}
```
AutoMapper сейчас не модная тема, он медленный, тк маппит через рефлексию + конфиги заебывает писать так же как мапить самому.
Вот видос со сравнением популярных мапперов: https://www.youtube.com/watch?v=U8gSdQN2jWI
На моих последних проектах вообще не юзают либы для маппинга и пишутся маппинги самостоятельно и в целом мне это нравится)

controlofthesea commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace TaskManager.Infrastructure.Repositories
+{
+    public class Repository<TEntity, TModels> : IRepository<TEntity, TModels> where TEntity : class where TModels : class
```
реализации сервисов инфрастктуры должны быть internal. По большому счету, все нужно стараться скрывать, а паблики в основном должны быть в коре

svirzhevskiy replied:
ты исправил только тут, а в остальных классах инфры оставил как и было - паблик)

controlofthesea commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -0,0 +1,60 @@
+            _mapper = mapper ?? throw new ArgumentException(nameof(mapper));
+        }
+
+        public async Task<IEnumerable<TModels>> GetAllAsync()
+        {
+            var entities = await _context.Set<TEntity>().ToListAsync();
+           
+            return _mapper.Map<IEnumerable<TModels>>(entities);
+        }
```
Если интересует, в автомаппере есть метод, который стыкается с ef и сразу маппит из базы в нужные модели, забыл как называется

controlofthesea replied:
вроде ProjectTo

controlofthesea commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -0,0 +1,60 @@
+        public async Task AddAsync(TModels dto)
+        {
+            var entity = _mapper.Map<TEntity>(dto);
+            _context.Set<TEntity>().Add(entity);
```
`_context.Set<TEntity>` можно вынести в приватную переменную класса, чтобы сократить код и не сетать в каждом методе тоже самое :)

controlofthesea commented:
По структуре мне нравится, так и представлял 🏌️‍♂️
По ef я уже почти все забыл сам :)

controlofthesea commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -0,0 +1,60 @@
+            return _mapper.Map<TModels>(entity);
+        }
+
+        public async Task AddAsync(TModels dto)
```
dto?

CrazyTapok replied:
Косяк, исправлю

CrazyTapok committed:
refactor: сorrect after review, replace AutoMapper by Mapperly

svirzhevskiy commented on src/TaskManager.Core/Interfaces/IRepository.cs:
```diff
@@ -0,0 +1,9 @@
+public interface IRepository<TEntity, TModel> where TEntity : class where TModel : class
+{
+    Task<IEnumerable<TModel>> GetAllAsync();
+    Task<TModel> GetByIdAsync(int id);
```
почему тут id int? не guid должен быть? и в delete

svirzhevskiy commented on src/TaskManager.Core/Enums/TaskStatus.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace TaskManager.Core.Enums;
```
сори за доеб, но хочется же пустую строку хотя бы между неймспейсом и классом :)

CrazyTapok replied:
Справедливо

svirzhevskiy commented on src/TaskManager.Infrastructure/EF/DBContext.cs:
```diff
@@ -0,0 +1,27 @@
+    public DbSet<Project> Projects { get; set; } = null!;
+    public DbSet<Task> Tasks { get; set; } = null!;
+
+    public DBContext(DbContextOptions<DBContext> options)
+        : base(options)
+    {
+        //Database.EnsureDeleted();
+        //Database.EnsureCreated();
+    }
```
а почему это закомментировано? я уже не помню про ef)

CrazyTapok replied:
для создания первой миграции.

svirzhevskiy replied:
не понимаю, когда это нужно разкомментить?

CrazyTapok replied:
что бы сделать первую миграцию сделать база должна быть чистой, т.е. не должно быть таблиц. поэтому и закомментил. теперь можно разкомментить.

svirzhevskiy commented on src/TaskManager.Infrastructure/EntityConfiguration/EmployeeConfigurayion.cs:
```diff
@@ -0,0 +1,38 @@
+using Microsoft.EntityFrameworkCore;
+using TaskManager.Infrastructure.Entities;
+
+namespace TaskManager.Infrastructure.EntityConfiguration;
+
+public class EmployeeConfigurayion : IEntityTypeConfiguration<Employee>
```
тут уже со строкой, красивее же)

svirzhevskiy commented on src/TaskManager.Infrastructure/EntityConfiguration/TaskConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+
+        builder.Property(t => t.CreatedDate)
+            .HasColumnType("datetime")
+            .HasDefaultValueSql("GETDATE()");
```
сейвовее хранить даты в UTC, а то потом не будешь знать что когда в каком поясе
GETUTCDATE()

svirzhevskiy commented on src/TaskManager.Infrastructure/Mappings/Mapper.cs:
```diff
@@ -0,0 +1,11 @@
+[Mapper]
+public static partial class Mapper
+{
+    public static partial T Map<K, T>(this K entity);
```
лично я не люблю когда любые переменные одной буквой называют T, K, i в for, x в лямдах - нихера не понятно же

CrazyTapok replied:
просто эта штука работает в обе стороны, все зависит как передашь параметры, она может model -> entity и entity -> model

svirzhevskiy replied:
почему тогда `this K entity`

svirzhevskiy commented on src/TaskManager.Infrastructure/Mappings/Mapper.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace TaskManager.Infrastructure.Mappings;
+
+[Mapper]
```
посмотрел видик?) что это за маппер?) лучший?)

CrazyTapok replied:
она на втором месте после manual

svirzhevskiy commented:
Насчет коммитов, ты уже пишешь что-то адекватное, что отражает то, что было сделано, но можно добавить структуры:
https://hackernoon.com/7-rules-for-writing-a-good-commit-message
Это первая ссылка с гугла, можешь еще сам поискать, почитать и выбрать, что понравится

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -0,0 +1,63 @@
+        await _context.SaveChangesAsync();
+    }
+
+    public async Task UpdateAsync(TModel model)
+    {
+        var entity = model.Map<TModel, TEntity>();
+        _context.Entry(entity).State = EntityState.Modified;
+       
+        await _context.SaveChangesAsync();
+    }
```
если у тебя в методе только один await, который ожидает результат в конце, можно не авейтить внутри метода а ретурнить Taskу, чтобы авейтил только вызывающий метод, по идее микробуст будет)

CrazyTapok committed:
refactor: fix remarks after the review

Add indents after namespace, replace typr int by Guid, change the UpdateAsync method in Repository, , create new migration

svirzhevskiy commented:
игноришь коммент, инкапсуляция важна
![image](https://github.com/user-attachments/assets/7eea6536-0eea-455c-b5bd-64370b852362)

svirzhevskiy commented on src/TaskManager.Infrastructure/Mappings/Mapper.cs:
```diff
@@ -0,0 +1,11 @@
+[Mapper]
+public static partial class Mapper
+{
+    public static partial TResult Map<TSource, TResult>(this TSource entity);
```
почему переменная называется entity?

CrazyTapok replied:
справил

CrazyTapok committed:
refactor: fix remarks after the review

change access modifiers, replace name variable

CrazyTapok committed:
refactor: fix remarks after the review

Delete mapper, delete entities in infrastructure, chenge IRepository and Repository

svirzhevskiy commented on src/TaskManager.Infrastructure/EntityConfiguration/EmployeeConfigurayion.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace TaskManager.Infrastructure.EntityConfiguration;
+
+internal class EmployeeConfigurayion : IEntityTypeConfiguration<Employee>
```
```suggestion
internal class EmployeeConfiguration : IEntityTypeConfiguration<Employee>
```

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -0,0 +1,51 @@
+        _dbSet = _context.Set<TModel>();
+    }
+
+    public Task<List<TModel>> GetAllAsync()
+    {
+        return _dbSet.ToListAsync();
+    }
+
+    public Task<TModel> GetByIdAsync(Guid id)
+    {
+        return _dbSet.FindAsync(id).AsTask();
+    }
+
+    public async Task AddAsync(TModel model)
+    {
+        await _dbSet.AddAsync(model);
+      
+        await _context.SaveChangesAsync();
+    }
```
Вроде все выглядит хорошо)
А теперь то, за что ты должен быть отъебан всегда - напиши юнит тесты хотя бы на репозиторий :)

svirzhevskiy commented on src/TaskManager.Infrastructure/EF/StartupInfrastructure.cs:
```diff
@@ -0,0 +1,12 @@
+
+public static class StartupInfrastructure
+{
+    public static void AddDbContextInfrastructure(this IServiceCollection services, IConfiguration configuration)
+        => services.AddDbContext<DBContext>(options =>
+            options.UseSqlServer(configuration.GetConnectionString("DefaultConnection")));
```
еще забыл тут подключить реализацию репозитория

CrazyTapok committed:
refactor: fix remarks after the review
Rename EmployeeConfiguration, connect repository implementation, add new project for tests, add RepositoryTests

svirzhevskiy commented on src/TaskManager.UnitTests/RepositoryTests.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace TaskManager.UnitTests
+{
+    public class RepositoryTests
+    {
+
+    }
+}
```
это лишний проект? удалить надо?

CrazyTapok replied:
хз, как это попало в git, в структуре проекта этого нет

svirzhevskiy replied:
значит оно не добавлено в sloution, но лежит на в папке

svirzhevskiy commented on src/TaskManager.xUnitTests/Infrastructure/RepositoryTests.cs:
```diff
@@ -0,0 +1,69 @@
+using Moq;
+using TaskManager.Core.Interfaces;
+
+namespace TaskManager.xUnitTests.Infrastructure;
```
почему помещаем название либы в название проекта? xUnitTests

svirzhevskiy replied:
достаточно просто Tests

CrazyTapok replied:
видел, как в туториалах так делают

svirzhevskiy replied:
это хуевый туториал, это точно. Потому что называть проект по названию либы - это антитуториал какой-то

svirzhevskiy commented on src/TaskManager.xUnitTests/Infrastructure/RepositoryTests.cs:
```diff
@@ -0,0 +1,69 @@
+    [Fact]
+    public async Task AddEntity_ShouldCallAddAsync()
+    {
+        var entity = new TestModel { Id = Guid.NewGuid(), Title = "Test 66" };
```
Прогугли либу Fixture, она позволяет создавать тестовые объекты автоматом, без ручеого заполнения полей.
Это очень полезно, когда, например, добавляешь в модель новое свойство, то при таком подходе тебе придется фиксить все тесты. С Fixture не придется

svirzhevskiy commented on src/TaskManager.xUnitTests/Infrastructure/RepositoryTests.cs:
```diff
@@ -0,0 +1,69 @@
+    }
+
+
+    public class TestModel
+    {
+        public Guid Id { get; set; }
+
+        public string Title { get; set; }
+    }
```
положи в отдельный файл или хотя бы приватной сделай

svirzhevskiy commented on src/TaskManager.xUnitTests/Infrastructure/RepositoryTests.cs:
```diff
@@ -0,0 +1,69 @@
+namespace TaskManager.xUnitTests.Infrastructure;
+public class RepositoryTests
+{
+    private readonly Mock<IRepository<TestModel>> _repository;
```
Всегда класс, который сейчас под тестами, должен создаваться настоящий с его настоящим конструктором. Когда ты мокаешь класс, который тестируешь, ты тестируешь мок - бред😊. Мокаются зависимости тестируемого класса, то есть здесь это будет мок датабейс

CrazyTapok committed:
refactor: fix remarks after the review
Add test project, add test for Repository, add TestModel, add Fixture lib, add TestDBContext

svirzhevskiy reviewed (approved):
Вот теперь хорошо!
